File: src/game/Vector.java
package game;

public class Vector {
	public static final Vector zeroVector = new Vector(0, 0);
	
	public final float x, y;
	
	public Vector(float x, float y) {
		this.x = x;
		this.y = y;
	}
	
	public Vector(float a) {
		this((float) Math.cos(a), (float) Math.sin(a));
	}
	
	public float angle() {
		return (float) Math.atan2(y, x);
	}
	
	public float magnitude() {
		return (float) Math.sqrt(x*x + y*y);
	}
	
	public Vector direction() {
		float mag = magnitude();
		if (mag == 0) {
			return zeroVector;
		} else {
			return this.scale(1/mag);
		}
	}
	
	public Vector scale(float s) {
		return new Vector(x*s, y*s);
	}
	
	public Vector negate() {
		return new Vector(-x, -y);
	}
	
	public Vector add(Vector v) {
		return new Vector(x + v.x, y + v.y);
	}
	
	public Vector sub(Vector v) {
		return this.add(v.negate());
	}
	
	public float dot(Vector v) {
		return x*v.x + y*v.y;
	}
	
	public Vector rotate(float a) {
		if (this.equals(zeroVector)) {
			return zeroVector;
		}
		
		return new Vector(a + angle()).scale(magnitude());
	}
	
	public float distance(Vector v) {
		return this.sub(v).magnitude();
	}
	
	public Vector perpCCW() {
		return new Vector(-y, x);
	}
	
	public Vector perpCW() {
		return new Vector(y, -x);
	}
	
	@Override
	public boolean equals(Object o) {
		try {
			Vector v = (Vector) o;
			return v.x == x && v.y == y;
		} catch (ClassCastException e) {
			e.printStackTrace();
		}
		return false;
	}
	
	@Override
	public int hashCode() {
		return Float.hashCode(x) ^ Float.hashCode(y);
	}
	
	@Override
	public String toString() {
		return "<" + x + ", " + y + ">";
	}
}

File: ts/game/Vector.ts
export class Vector {
  static readonly zeroVector = new Vector(0, 0);

  readonly x: number;
  readonly y: number;

  constructor(x: number, y?: number) {
    if (y === undefined) {
      this.x = Math.cos(x);
      this.y = Math.sin(x);
    } else {
      this.x = x;
      this.y = y;
    }
  }

  angle(): number {
    return Math.atan2(this.y, this.x);
  }

  magnitude(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  direction(): Vector {
    const mag = this.magnitude();
    if (mag === 0) {
      return Vector.zeroVector;
    }
    return this.scale(1 / mag);
  }

  scale(s: number): Vector {
    return new Vector(this.x * s, this.y * s);
  }

  negate(): Vector {
    return new Vector(-this.x, -this.y);
  }

  add(v: Vector): Vector {
    return new Vector(this.x + v.x, this.y + v.y);
  }

  sub(v: Vector): Vector {
    return this.add(v.negate());
  }

  dot(v: Vector): number {
    return this.x * v.x + this.y * v.y;
  }

  rotate(a: number): Vector {
    if (this.equals(Vector.zeroVector)) {
      return Vector.zeroVector;
    }
    return new Vector(a + this.angle()).scale(this.magnitude());
  }

  distance(v: Vector): number {
    return this.sub(v).magnitude();
  }

  perpCCW(): Vector {
    return new Vector(-this.y, this.x);
  }

  perpCW(): Vector {
    return new Vector(this.y, -this.x);
  }

  equals(o: unknown): boolean {
    if (!(o instanceof Vector)) return false;
    return o.x === this.x && o.y === this.y;
  }

  toString(): string {
    return `<${this.x}, ${this.y}>`;
  }
}
